### DISCUSSION TITLE
Is terminal.buffer._core.viewport._activeBuffer.x the right way to access the current cursor x position

### DISCUSSION THREAD

**Discussion by ckkMicrosoft** [sourcelair/xterm.js#4392] [Q&A] [answered]
This is a wonderful library :) .. I love how easy it was for me to get up and running! I'd really appreciate some help here.

The examples on the net indicate that I should be able to use term.buffer.cursorX to get the current x position but it seems that the latest code has changed and its no longer supported. I was able to access the x position by trial and error but the underscores make me think that this may not be the right way.

Is there a right way to get the cursor position or has there been a regression?

**Comment by Tyriar** [comment, marked as answer]
As an embedder you should never reach into anything not specified in xterm.d.ts like `_core`, as that's getting into private API that can change at any time. The right way to access the cursor x is `term.buffer.active.cursorX`. The active is needed because there is a "normal" and an "alt" buffer (eg. when in vim).

**Reply by ckkashyap** [reply]
Thank you so much!

**Reply by alanhe421** [reply]
Is there a way to obtain the width of a cell? What I ultimately want to achieve is the relative position of the cursor within the entire terminal, such as how many pixels the x-coordinate and y-coordinate are. I have a requirement where a pop-up window appears at the cursor's position based on a certain hotkey."

**Reply by Tyriar** [reply]
@alanhg `term.buffer.active.getLine(y)?.getCell(x)?.getWidth()`

**Reply by alanhe421** [reply]
This is the width of a single cell. Is there a way for me to obtain the relative position of the cursor in the terminal, that is, the relative width and height?

**Reply by Tyriar** [reply]
If you want by pixels, I don't think there's an easy way to get that. You could use the DOM renderer and get the cursor element's bounding box, you would lose the perf benefits though.

**Reply by alanhe421** [reply]
So here, for my needs, I can only calculate the width and height of each cell and sum them up for use. Indeed, the performance would be poor.

**Reply by jerch** [reply]
Well you can try to get somewhere with the private API here: https://github.com/xtermjs/xterm.js/blob/823258e31150d47bc79283506c377d32686975e1/src/browser/services/RenderService.ts#L53

It holds more information, but ofc might change from version to version, so creates a much higher maintenance burden on your side.

Edit:
It still would need you to calc offsets for individual cell access/positioning, as the terminal does not maintain any data structure containing that precalced.
